great_expectations/execution_engine/split_and_sample/sparkdf_data_splitter.py

```diff
@@ -0,0 +1,288 @@
+            column_name=column_name,
+            batch_identifiers=batch_identifiers,
+            date_parts=[DatePart.YEAR, DatePart.MONTH],
+        )
+
+    def split_on_year_and_month_and_day(
+        self,
+        df: DataFrame,
+        column_name: str,
+        batch_identifiers: dict,
+    ) -> DataFrame:
+        """Split on year and month and day values in column_name.
+
+        Args:
+            df: dataframe from batch data.
+            column_name: column in table to use in determining split.
+            batch_identifiers: should contain a dateutil parseable datetime whose
+                relevant date parts will be used for splitting or key values
+                of {date_part: date_part_value}.
+
+        Returns:
+            List of boolean clauses based on whether the date_part value in the
+                batch identifier matches the date_part value in the column_name column.
+        """
+        return self.split_on_date_parts(
+            df=df,
+            column_name=column_name,
+            batch_identifiers=batch_identifiers,
+            date_parts=[DatePart.YEAR, DatePart.MONTH, DatePart.DAY],
+        )
+
+    def split_on_date_parts(
+        self,
+        df: DataFrame,
+        column_name: str,
+        batch_identifiers: dict,
+        date_parts: Union[List[DatePart], List[str]],
+    ) -> DataFrame:
+        """Split on date_part values in column_name.
+
+        Values are NOT truncated, for example this will return data for a
+        given month (if only month is chosen for date_parts) for ALL years.
+        This may be useful for viewing seasonality, but you can also specify
+        multiple date_parts to achieve date_trunc like behavior e.g.
+        year, month and day.
+
+        Args:
+            df: dataframe from batch data.
+            column_name: column in data used to determine split.
+            batch_identifiers: should contain a dateutil parseable datetime whose date parts
+                will be used for splitting or key values of {date_part: date_part_value}
+            date_parts: part of the date to be used for splitting e.g.
+                DatePart.DAY or the case-insensitive string representation "day"
+
+        Returns:
+            Dataframe with splitting applied.
+        """
+        self._validate_date_parts(date_parts)
+
+        date_parts: List[DatePart] = self._convert_date_parts(date_parts)
+
+        column_batch_identifiers: dict = batch_identifiers[column_name]
+
+        date_parts_dict: dict = (
+            self._convert_datetime_batch_identifiers_to_date_parts_dict(
+                column_batch_identifiers, date_parts
+            )
+        )
+
+        for date_part, date_part_value in date_parts_dict.items():
+            df = df.filter(
+                getattr(F, self._convert_date_part_to_spark_equivalent(date_part))(
+                    F.col(column_name)
+                )
+                == date_part_value
+            )
```

Comment:
Just for my understanding, this is updating the existing muatble df over and over again? The return obj on L161 recognizes all these filters?

Reply:
That is correct, it's as if I'm AND-ing all of the conditions together. They are added to the dataframe lazily and then executed when we call .collect() later when resolving a metric. I'm relying on the spark executor to optimize compute. If we find performance issues, another option is to build up a query with all of the conditions - but I think spark will treat those cases equivalently.